towhee/hub/downloader.py

```diff
@@ -0,0 +1,178 @@
+        files_dir.mkdir(parents=True, exist_ok=True)
+        return files_dir
+
+    def get_tag_path(self, create = False):
+        versions_dir = self._root / 'versions' / self._tag
+        if create:
+            versions_dir.mkdir(parents=True, exist_ok=True)
+        return versions_dir
+
+    @property
+    def requirements(self):
+        for item in self._meta_infos:
+            if item['type'] == 'blob' and item['path'] == 'requirements.txt':
+                return self.get_tag_path() / item['path']
+        return None
+
+    def symlink_files(self):
+        tmp_dir = Path(tempfile.mkdtemp(dir=self._root))
+        for dst, src in self.symlink_pair():
+            dst_file = tmp_dir / dst
+            dst_file.parent.mkdir(parents=True, exist_ok=True)
+            dst_file.symlink_to(src)
+        tmp_dir.rename(self.get_tag_path(True))
+
+    def symlink_pair(self):
+        pair = []
+        for item in self._meta_infos:
+            if item['type'] == 'tree':
+                continue
+            pair.append((Path(item['path']), self.file_path / item['sha']))
+        return pair
+
+    def local_url_pair(self):
+        pair = []
+        for item in self._meta_infos:
+            if item['type'] == 'tree':
+                continue
+            pair.append((self.file_path / item['sha'], item['downloadLink']))
+        return pair
+
+
+class _Downloader:
+    """
+    Download op
+    """
+    def __init__(self, hub_files: 'HubFiles'):
+        self._hub_files = hub_files
+
+    @staticmethod
+    def download_url_to_file(url: str, dst: str, file_size: int = None) -> bool:
+        """Download file at the given URL to a local path.
+        Args:
+            url (str): URL of the file to download
+            dst (str): Full path where file will be saved
+        """
+
+        with requests.get(url, stream=True) as r:
+            r.raise_for_status()
+            try:
+                if file_size is None:
+                    file_size = int(r.headers.get('content-length', 0))
+                dst_dir = os.path.dirname(dst)
+                with tempfile.NamedTemporaryFile(delete=False, dir=dst_dir) as f:
+                    with tqdm(total=file_size, unit='B', unit_scale=True, desc=f'Downloading {url} to {dst_dir}') as pbar:
+                        for chunk in r.iter_content(chunk_size=8192):
+                            real_size = f.write(chunk)
+                            pbar.update(real_size)
+                shutil.move(f.name, dst)
+            finally:
+                if os.path.exists(f.name):
+                    os.remove(f.name)
+
+    def download(self):
+        futures = []
+        with ThreadPoolExecutor(max_workers=5) as pool:
+            for local_file, url in self._hub_files.local_url_pair():
+                if not local_file.is_file():
+                    futures.append(pool.submit(_Downloader.download_url_to_file, url, local_file))
+            _ = [i.result() for i in futures]
```

Comment:
exception handling?

Reply:
This function would raise the exception, and the `operator_loader` will catch the exception, print logs and stop running.